### PR TITLE
Plugins: Enable qmake for core22

### PR DESCRIPTION
**Pull request by ScarlettGatelyMoore**
- [x ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `make lint`?
- [ ] Have you successfully run `pytest tests/unit`?

-----

### CODE REVIEW

**Comment by lengau** [comment]
This seems to me like it would make sense upstream as a [craft-parts plugin](https://github.com/canonical/craft-parts/tree/main/craft_parts/plugins), since I can imagine wanting to use qmake for charms or rocks as well.

File: tests/spread/plugins/craft-parts/build-and-run-hello/qmake-hello/snap/snapcraft.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+name: qmake-hello
+version: "1.0"
+summary: test the qmake plugin
+description: |
+  This is a basic qmake snap. It just prints a hello world.
+  If you want to add other functionalities to this snap, please don't.
+  Make a new one.
+
+grade: devel
+base: core20
```
Comment (lengau) [inline tests/spread/plugins/craft-parts/build-and-run-hello/qmake-hello/snap/snapcraft.yaml:10]:
```suggestion
base: core22
```

File: snapcraft/parts/plugins/qmake.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      the qmake project file to use. This is usually only needed if
+      qmake can not determine what project file to use on its own.
+"""
+
+from typing import Any, Dict, List, Set
+
+from craft_parts import plugins
+
+
+class QMakePlugin(qmake.QmakePlugin):
```
Comment (lengau) [inline snapcraft/parts/plugins/qmake.py:42]:
```suggestion
class QMakePlugin(plugins.Plugin):
```

File: snapcraft/parts/plugins/qmake.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            "properties": {
+                "qmake-parameters": {
+                    "type": "array",
+                    "uniqueItems": True,
+                    "items": {"type": "string"},
+                    "default": [],
+                },
+                "qmake-project-file": {"type": "string", "default": ""},
+            },
+        }
```
Comment (lengau) [inline snapcraft/parts/plugins/qmake.py:58]:
```suggestion
```

This is no longer used in `craft-parts` plugins :-)

File: snapcraft/parts/plugins/qmake.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        return " ".join(cmd)
+
+    def get_build_commands(self) -> List[str]:
+        return [
+            self._get_qmake_configure_command(),
+            # Avoid overriding the CFLAGS and CXXFLAGS environment
+            # variables qmake sets in the generated Makefile
+            'env -u CFLAGS -u CXXFLAGS make -j"${CRAFT_PARALLEL_BUILD_COUNT}"',
+            'make install INSTALL_ROOT="${CRAFT_PART_INSTALL}"',
```
Comment (lengau) [inline snapcraft/parts/plugins/qmake.py:98]:
```suggestion
            f'env -u CFLAGS -u CXXFLAGS make "-j{self._part_info.parallel_build_count}"',
            f'make install INSTALL_ROOT="{self._part_info.part_install_dir}"',
```

File: snapcraft/parts/plugins/qmake.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    def out_of_source_build(self):
+        return True
+
+    def _get_qmake_configure_command(self) -> str:
+        cmd = [
+            "qmake",
+            'QMAKE_CFLAGS+="${CFLAGS:-}"',
+            'QMAKE_CXXFLAGS+="${CXXFLAGS:-}"',
+            'QMAKE_LFLAGS+="${LDFLAGS:-}"',
+        ] + self.options.qmake_parameters
```
Comment (lengau) [inline snapcraft/parts/plugins/qmake.py:79]:
```suggestion
        ] + self._options.qmake_parameters
```

File: snapcraft/parts/plugins/qmake.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+        if self.options.qmake_project_file:
```
Comment (lengau) [inline snapcraft/parts/plugins/qmake.py:81]:
```suggestion
        if self._options.qmake_project_file:
```

File: snapcraft/parts/plugins/qmake.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            cmd.append(
+                '"${{CRAFT_PART_SRC_WORK}}/{}"'.format(
+                    self.options.qmake_project_file
+                )
+            )
```
Comment (lengau) [inline snapcraft/parts/plugins/qmake.py:86]:
```suggestion
            cmd.append(
                f'"{self._part_info.part_src_dir}/{self._options.qmake_project_file}"'
            )
```

File: snapcraft/parts/plugins/qmake.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        else:
+            cmd.append('"${CRAFT_PART_SRC_WORK}"')
```
Comment (lengau) [inline snapcraft/parts/plugins/qmake.py:88]:
```suggestion
            cmd.append(self._part_info.part_build_dir)
```

**Comment by ScarlettGatelyMoore** [comment]
> This seems to me like it would make sense upstream as a [craft-parts plugin](https://github.com/canonical/craft-parts/tree/main/craft_parts/plugins), since I can imagine wanting to use qmake for charms or rocks as well.

Thank you so much for the review. All changes applied!

**Comment by lengau** [comment]
Thanks for this, @ScarlettGatelyMoore ! It looks great, and I especially appreciate the spread test.

I spoke with @sergiusens this morning and he confirmed we should be putting the plugin itself into [craft-parts](https://github.com/canonical/craft-parts/tree/main/craft_parts/plugins). Could you move the Python code into a PR on `craft-parts`, but keep this PR open with the spread tests? That way we'll have upstream'd `qmake` support, but with a full e2e test here.

**Comment by icotoi** [comment]
Any updates when this can be merged?

**Comment by patviafore-kudzera** [comment]
I'd also appreciate a merge, as I'm running into this issue as well. Can I help with anything that's blocking that people don't have time to dig into?

**Comment by lengau** [comment]
Status report on this:

We need a release of `craft-parts` to include [this PR](https://github.com/canonical/craft-parts/pull/463). Then presuming the tests succeed (I don't see why they wouldn't) we can merge this.

@cmatsuoka how do you feel about a craft-parts 1.27 release? It looks like we'll get the qmake plugin, proxy support for ant, and using the `rustup` snap.

**Review by mr-cal — APPROVED** [review]

**Review by lengau — APPROVED** [review]

**Pull request merged**